src/main/kotlin/org/elm/workspace/ElmFormatCLI.kt

```diff
@@ -1,27 +1,67 @@
 package org.elm.workspace
 
+import com.intellij.execution.ExecutionException
 import com.intellij.execution.process.ProcessOutput
-import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.vfs.VirtualFile
+import org.elm.lang.core.psi.ElmFile
 import org.elm.openapiext.GeneralCommandLine
 import org.elm.openapiext.execute
-import org.elm.openapiext.pathAsPath
-import org.elm.openapiext.withWorkDirectory
+import org.elm.openapiext.isSuccess
+import org.elm.openapiext.toPsiFile
 import java.nio.file.Path
 
 class ElmFormatCLI(val elmFormatExecutablePath: Path) {
 
-    fun reformatFile(project: Project, elmVersion: Version, file: VirtualFile, owner: Disposable = project): ProcessOutput {
+    private fun getFormattedContentOfDocument(elmVersion: Version, document: Document): ProcessOutput {
         val arguments = listOf(
                 "--yes",
-                "--elm-version",
-                "${elmVersion.x}.${elmVersion.y}",
-                file.path
+                "--elm-version=${elmVersion.x}.${elmVersion.y}",
+                "--stdin"
         )
+
         return GeneralCommandLine(elmFormatExecutablePath)
-                .withWorkDirectory(file.parent.pathAsPath)
                 .withParameters(arguments)
-                .execute(owner)
+                .execute(document.text)
+    }
+
+    fun formatDocumentAndSetText(project: Project, document: Document, version: Version) {
+
+        val result = ProgressManager.getInstance().runProcessWithProgressSynchronously<ProcessOutput, ExecutionException>({
+            getFormattedContentOfDocument(version, document)
+        }, "Running elm-format on current file...", true, project)
+
+        if (result.isSuccess) {
+            val formatted = result.stdout
+            val source = document.text
+
+            if (source != formatted) {
+                ApplicationManager.getApplication().runWriteAction {
```

Comment:
While we will want the undo-transparent action later when this is invoked via save hook, right now we want to add to the undo stack. I know you said that you tried and ran into problems. I gave it a go and had problems too, but I eventually figured it out. You have to call `CommandProcessor#executeCommand` and do the document update inside the runnable that you give it.

Here's the solution that works:
```kotlin
        if (result.isSuccess) {
            val formatted = result.stdout
            if (document.text != formatted) {
                CommandProcessor.getInstance().executeCommand(project, {
                    ApplicationManager.getApplication().runWriteAction {
                        document.setText(formatted)
                    }
                }, "apply elm-format output", null, document)
            }
        }
```

I also made it so that we wait to acquire the write action until as late as possible--this follows what I've seen in other usages of `executeCommand` from within the IntelliJ codebase.

Reply:
Oh, how cool that you figured it out. 👍 I wish Jetbrains plugin documentation would be a bit better.

Reply:
I tried the new behaviour with my current impl of the save hook. It works as expected and there is no "undo reload from disk". :)